Add NVIDIA workload test definition

stmcginnis opened the pull request:
**Issue number:**

Related: #419

**Description of changes:**

This adds a new `tests` directory tree to contain our workload tests and any other type of add-on test type we decide to add in the future. It adds our first workload test for running NVIDIA smoketests. Basic instructions are included to help guide future test authors of the requirements for tests and where they should be placed.

Some of this content will need to be updated once we fully finish publishing the container images and flesh out things like ECS testing and how the full execution of workload tests will be performed. This gets the majority of the pieces in place so those updates will be smaller and it will be easier to complete that work using published content.

**Testing done:**

Most testing was done when the initial `k8s-workload-agent` was added. The `nvidia-smoke` container creation has been tested by building the multiarch image and running it manually on both amd64 and arm64 ECS Bottlerocket nodes and viewing the output results.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

stmcginnis committed:
Add NVIDIA workload test definition

This adds a new `tests` directory tree to contain our workload tests and
any other type of add-on test type we decide to add in the future. It
adds our first workload test for running NVIDIA smoketests. Basic
instructions are included to help guide future test authors of the
requirements for tests and where they should be placed.

Some of this content will need to be updated once we fully finish
publishing the container images and flesh out things like ECS testing
and how the full execution of workload tests will be performed. This
gets the majority of the pieces in place so those updates will be
smaller and it will be easier to complete that work using published
content.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

ecpullen commented on tests/workload/nvidia-smoke/Makefile:
```diff
@@ -0,0 +1,25 @@
+.DEFAULT_GOAL:=publish
+
+#### Variables
+REGISTRY ?=
```
Should we use the same name as the root makefile `PUBLISH_IMAGES_REGISTRY`?

stmcginnis replied:
Yeah, that makes sense. I don't really like the length of that variable name, but it's better to have consistency than less typing.

ecpullen commented on tests/workload/nvidia-smoke/Makefile:
```diff
@@ -0,0 +1,25 @@
+REPOSITORY ?= gpu-tests
+
+#### Sanity check we have everything
+ifndef REGISTRY
+$(error "The REGISTRY value must be provided (e.g. REGISTRY=861807767978.dkr.ecr.us-east-2.amazonaws.com make)")
+endif
```
Is there a reason a user may not need to specify the `REGISTRY` i.e `kind load`

stmcginnis replied:
Probably not in the case of this particular workload test, but I see what you're saying. Since this is setting the pattern for future workload tests, I'll update this to check if the variable is set. If so, then it can format the `$REGISTRY/` prefix, and if not it can be empty.

ecpullen commented on tests/workload/nvidia-smoke/README.md:
```diff
@@ -0,0 +1,90 @@
+Compiling the `amd64` binaries on an `arm64` host using emulation will be significantly slower.
+See below for instructions on how to set up a remote build host to offload this activity.
+
+This guide assumes you use ECR as your container's registry, and that you already have the ECR credentials plugin configured in your host.
```
Is it possible to write this guide without assuming `ECR`

stmcginnis replied:
Reworded this note since there really isn't anything that assumes or expects ECR vs any other image registry.

ecpullen reviewed (commented):
Awesome!

stmcginnis commented:
I am now seeing the same failures on arm64 as I was seeing when trying to build on amd64:

```txt
 > [linux/amd64 builder  5/15] RUN cd /cuda-samples/Samples/0_Introduction/vectorAdd/ && make && cp $(basename $(pwd)) /samples/:
#0 0.200 Makefile:86: *** ERROR - unsupported value Error while loading /usr/local/nvidia/bin/uname: No such file or directory for TARGET_ARCH!.  Stop.
------
Dockerfile:21
--------------------
  19 |     # it filters out what samples won't be build.
  20 |     # This only includes the samples that should run on both amd64 and arm64.
  21 | >>> RUN cd /cuda-samples/Samples/0_Introduction/vectorAdd/ && make && cp $(basename $(pwd)) /samples/
  22 |     RUN cd /cuda-samples/Samples/0_Introduction/simpleVoteIntrinsics/ && make && cp $(basename $(pwd)) /samples/
  23 |     RUN cd /cuda-samples/Samples/0_Introduction/simpleAtomicIntrinsics/ && make && cp $(basename $(pwd)) /samples/
--------------------
ERROR: failed to solve: process "/dev/.buildkit_qemu_emulator /bin/sh -c cd /cuda-samples/Samples/0_Introduction/vectorAdd/ && make && cp $(basename $(pwd)) /samples/" did not complete successfully: exit code: 2
make: *** [Makefile:17: build] Error 1
```

Going to mark as draft until I understand the root cause and what has changed from the 50+ times I had built this successfully before submitting this PR.

webern commented on tests/workload/README.md:
```diff
@@ -0,0 +1,77 @@
+results_dir="${RESULTS_DIR:-/tmp/results}"
+mkdir -p "${results_dir}"
+
+# Example of option argument passed as environment variable to the container
```
```suggestion
# Example of optional argument passed as environment variable to the container
```

stmcginnis commented:
Had to take a slightly different approach to be able to get a consistent good build. It has the added benefit that it is much, much faster due to not needing to compile under emulation, so the build completes in a fraction of the time from the first approach.

Addressed comments and moved the test tree under `bottlerocket/` since these are all Bottlerocket specific things.

webern commented on bottlerocket/tests/workload/README.md:
```diff
@@ -0,0 +1,77 @@
+
+# Make sure to always capture results in expected place and format
+trap saveResults EXIT
+
+/bin/tests/benchmark --name "${name}" >> "${results_dir}/${test_name}-results" 2>&1
```
```suggestion
# Run whatever you intend to test using this container.
/bin/tests/benchmark --name "${name}" >> "${results_dir}/${test_name}-results" 2>&1
```

webern commented on bottlerocket/tests/workload/nvidia-smoke.yaml:
```diff
@@ -0,0 +1,16 @@
+spec:
+  agent:
+    name: workload-agent
+    image: __WORKLOAD_AGENT_IMAGE__
```
Is this a templating mechanism?

stmcginnis replied:
I was trying to strike a balance between showing an example that someone could read in the docs and understanding what they would need to put in a file, and having something that we could potentially do some sort of sed replacement to use it as a template and pull it into some of our testing workflow.

This could be changed to have some dummy value if you think this is too odd or confusing. We could also wait until the official images are published and update these with those values once they are available.

ecpullen replied:
In most of the documentation I have written I have used `<SCREAMING-KEBAB-CASE>` to identify a value that should be replaced.

stmcginnis replied:
I've changed this to match that pattern.

webern commented on bottlerocket/tests/workload/nvidia-smoke/Makefile:
```diff
@@ -0,0 +1,28 @@
+publish:
+	docker buildx build --push --platform linux/arm64,linux/amd64 --tag $(REGISTRY)$(REPOSITORY):$(TAG) .
+
+x86_64:
+	docker buildx build --platform linux/amd64 --tag $(REGISTRY)$(REPOSITORY):$(TAG)-x86_64 .
+
+aarch64:
+	docker buildx build --platform linux/arm64 --tag $(REGISTRY)$(REPOSITORY):$(TAG)-aarch64 .
```
Should we have a multiarch manifest instead?

stmcginnis replied:
These last two are really just convenience targets and could really be removed. They were useful when testing things out on different hosts. In practice, the publish target is really all that is needed and it does create a multiarch image.

webern commented on bottlerocket/tests/workload/nvidia-smoke/README.md:
````diff
@@ -0,0 +1,68 @@
+You will need to run the following to create a new builder using the `docker-container` driver:
+
+```bash
+docker buildx create --use --bootstrap
````
This environment set up and subsequent 15 steps make me sad. :cry: 


Do I understand correctly that you are running another machine for the target architecture? Is it common to do with qemu?
Not your fault though, cross-arch building of containers is what it is...

Thank you for documenting this well.

stmcginnis replied:
Yeah, that's how I originally had it - with one host. It made cross-compiling incredibly slow.

And it was working for me running from an arm64 host fine, but when I tried running from an amd64 host I would get weird compile/linking errors with the NVIDIA sample binaries. At first I chalked this up to something odd with building for ARM on AMD, but then after cleaning up and resetting things I started to get the same errors on my arm64 host that had been previously working.

So I think in general it is fine to do both from one host, but in this particular case there is something about how the Makefiles do various checks and set variables that control compilation that somehow gets messed up when using the normal QEMU based route.

Unfortunately the only way I could find to reliably get things to build right was the currently documented approach of using two different machines. I'm not happy about it, but it's the best I could come up with.

It does have the nice benefit that compilation is much faster than the previous emulated approach - by probably 5x. So there is some setup pain to do it this way, but it kind of pays off in other ways.
